### PR TITLE
Remove call to proto.Clone() in http2Server.WriteStatus.

### PR DESCRIPTION
**Pull request by easwars**
Fixes https://github.com/grpc/grpc-go/issues/2182

* Expose a method from the internal package to get to the raw
  StatusProto wrapped by the status error, and use it from
  http2Server.WriteStatus().
* Add a helper method in internal/testutils to compare two status errors
  and update test code to use that instead of reflect.DeepEqual()

### CODE REVIEW

**Comment by dfawley** [comment]
- The `go.mod` and `go.sum` changes should be reverted from this PR.
- Consider squashing all your commits down to 1 (or a small number if there are relevant discrete changes for one PR) before sending the first iteration of a PR.  You can do this now and then `git push -f` your branch.
- For best results, use `git fetch upstream master && git rebase upstream/master` to pull mainline commits into your branch when needed instead of using `git merge`.

**Comment by easwars** [comment]
> * The `go.mod` and `go.sum` changes should be reverted from this PR.
This is done now.

> * Consider squashing all your commits down to 1 (or a small number if there are relevant discrete changes for one PR) before sending the first iteration of a PR.  You can do this now and then `git push -f` your branch.
I think I have squashed my three commits made as part of this PR now. From next time, will do a better job.

> * For best results, use `git fetch upstream master && git rebase upstream/master` to pull mainline commits into your branch when needed instead of using `git merge`.
I'm guessing, this is for the future.

Thanks for the git tips. Definitely going to be useful moving forward.

**Comment by easwars** [comment]
> * The `go.mod` and `go.sum` changes should be reverted from this PR.
> >   This is done now.

vet is failing because the reverted go.sum and go.mod changes? How do we usually handle this?

> * Consider squashing all your commits down to 1 (or a small number if there are relevant discrete changes for one PR) before sending the first iteration of a PR.  You can do this now and then `git push -f` your branch.
> >  I think I have squashed my three commits made as part of this PR now. From next time, will do a better job.
> 
> * For best results, use `git fetch upstream master && git rebase upstream/master` to pull mainline commits into your branch when needed instead of using `git merge`.
> >  I'm guessing, this is for the future.
> 

Thanks for the git tips. Definitely going to be useful moving forward.

File: internal/transport/http2_server.go [hunk 1]
```diff
@@ -817,7 +819,8 @@ func (t *http2Server) WriteStatus(s *Stream, st *status.Status) error {
 	headerFields = append(headerFields, hpack.HeaderField{Name: "grpc-status", Value: strconv.Itoa(int(st.Code()))})
 	headerFields = append(headerFields, hpack.HeaderField{Name: "grpc-message", Value: encodeGrpcMessage(st.Message())})
 
-	if p := st.Proto(); p != nil && len(p.Details) > 0 {
+	srp := internal.StatusRawProto.(func(*status.Status) *spb.Status)
```
[the remark this answers is not in the document]
Comment (easwars) [reply internal/transport/http2_server.go]: Done. Thanks.

**Comment by dfawley** [comment]
I also meant to request some performance metrics for this change - can you run a QPS-style workload (many concurrent, small RPCs) before/after and see if this makes much difference in performance / allocations?  Benchmark results should be posted in the first PR comment.  Thanks.

**Comment by easwars** [comment]
I ran a benchmark with 100 concurrent calls and req and resp sizes of 32 bytes. There doesn't seem to any significant change in performance.

Stream-traceMode_false-latency_0s-kbps_0-MTU_0-maxConcurrentCalls_100-reqSize_32B-respSize_32B-Compressor_off-Preloader_false
     Title       Before        After Percentage
  Bytes/op         1345         1346     0.07%
 Allocs/op           32           32     0.00%
50 latency    223.788µs    227.003µs     1.44%
90 latency    288.836µs    290.117µs     0.44%

**Comment by dfawley** [comment]
Interesting. Do we trust the allocation numbers?  I would expect them to be lower (both count and size) since we are removing `proto.Clone` calls.  Though the status protos are very small without any "details" additions, so I would expect very small gains.  Would you mind modifying the benchmarks locally to include a large details message in the server's response status to see if we can coerce a noticeable performance delta out of the benchmarks?

**Comment by easwars** [comment]
I made a change to the server to return a status with details message.

When I used a details message of len 128, this was the result:
     Title       Before        After Percentage
  Bytes/op        14275        13979    -2.07%
 Allocs/op          170          166    -2.35%


And when I used a details message of len 512, this was the result:
     Title       Before        After Percentage
  Bytes/op        17961        17281    -3.79%
 Allocs/op          172          168    -2.33%

The allocation numbers seem to be reasonably trustworthy in the sense that the number of allocs/op has not changed based on the length of the details message while the bytes/op has (and is proportional in some sense to the length of the details message).

**Comment by easwars** [comment]
@dfawley : Is there a way to get the vet test to be happy without the go.mod and go.sum changes?

File: internal/testutils/status_equal.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package testutils
+
+import (
+	"github.com/gogo/protobuf/proto"
```
Comment (dfawley) [inline internal/testutils/status_equal.go]: This should not use gogo/proto

Comment (easwars) [reply internal/testutils/status_equal.go]: Done. Thanks.
Some of the tests are unusually flaky today. Unable to make travis happy as of now.

**Pull request merged**